src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetInfoService.cs

```diff
@@ -44,25 +45,42 @@ internal abstract class AbstractSnippetInfoService : ForegroundThreadAffinitized
 
         private readonly IAsynchronousOperationListener _waiter;
 
+        /// <summary>
+        /// Collection of tasks started to populate the snippets cache.
+        /// This ensures we wait for them during shutdown.
+        /// </summary>
+        private readonly JoinableTaskCollection _asyncTasks;
+
         public AbstractSnippetInfoService(
             IThreadingContext threadingContext,
             Shell.SVsServiceProvider serviceProvider,
             Guid languageGuidForSnippets,
             IAsynchronousOperationListenerProvider listenerProvider)
             : base(threadingContext)
         {
-            AssertIsForeground();
+            _asyncTasks = new JoinableTaskCollection(threadingContext.JoinableTaskContext);
+            _waiter = listenerProvider.GetListener(FeatureAttribute.Snippets);
+            _languageGuidForSnippets = languageGuidForSnippets;
 
             if (serviceProvider != null)
             {
-                var textManager = (IVsTextManager2)serviceProvider.GetService(typeof(SVsTextManager));
-                if (textManager.GetExpansionManager(out _expansionManager) == VSConstants.S_OK)
+                _asyncTasks.Add(threadingContext.JoinableTaskFactory.RunAsync(() =>
                 {
-                    ComEventSink.Advise<IVsExpansionEvents>(_expansionManager, this);
-                    _waiter = listenerProvider.GetListener(FeatureAttribute.Snippets);
-                    _languageGuidForSnippets = languageGuidForSnippets;
-                    PopulateSnippetCaches();
-                }
+                    return InitializeAndPopulateSnippetsCacheAsync(threadingContext, serviceProvider);
+                }));
+            }
+        }
+
+        private async Task InitializeAndPopulateSnippetsCacheAsync(IThreadingContext threadingContext, Shell.SVsServiceProvider serviceProvider)
+        {
+            await threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
+            var asyncProvider = (Shell.IAsyncServiceProvider)serviceProvider;
```

Comment:
Consider updating this type to accept the serviceProvider as an IAsyncServiceProvider directly rather than having the cast down here. You may still have to do the cast but at least the requirements of this class are stated at the boundary.

Reply:
I'd rather leave it down here - I have to switch to the main thread to cast it.  If I move it up I have to switch before calling this method, but I'd rather leave all the messiness inside it.

Reply:
Casting things off the UI thread that came from a service provider is a no-no, I'm not aware of any rule about casting the serviceProvider _itself_. Or at least, we're doing this in one place:

https://github.com/dotnet/roslyn/blob/ffb3ca784e92b633f54dcc50f5306aaf945af3d0/src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcherProvider.cs#L25

So if that's a no-no, uh, it's not new. :smile:

Reply:
I can't use [Import(typeof(SVsServiceProvider))] the snippetinfo is using System.Composition for ExportLanguageService - so I just cast it before passing in.